crates/math/src/arith_expr.rs

```diff
@@ -441,6 +615,226 @@ impl<F: Field> Product for ArithExpr<F> {
 	}
 }
 
+impl<F: Field> SerializeBytes for ArithExpr<F> {
+	fn serialize(
+		&self,
+		write_buf: impl BufMut,
+		mode: SerializationMode,
+	) -> Result<(), SerializationError> {
+		/// Return the upper bound on the number of steps needed to serialize the expression.
+		fn estimate_steps<F: Field>(expr: &ArithExpr<F>) -> usize {
+			match expr {
+				ArithExpr::Const(_) => 1,
+				ArithExpr::Var(_) => 0, // variables do not represent a separate step
+				ArithExpr::Add(left, right) | ArithExpr::Mul(left, right) => {
+					1 + estimate_steps(left) + estimate_steps(right)
+				}
+				ArithExpr::Pow(base, _) => 1 + estimate_steps(base),
+			}
+		}
+
+		let steps_bound = estimate_steps(self);
+		stackalloc_with_default::<ArithCircuitStep<F>, _, _>(steps_bound, |mut steps_buf| {
```

Comment:
Please just use a Vec. This isn't performance critical, and we should only use `stackalloc` where absolutely necessary.